src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/ContentCulturePickerController.cs

```diff
@@ -0,0 +1,75 @@
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using OrchardCore.ContentManagement;
+using Microsoft.AspNetCore.Localization;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Localization;
+using OrchardCore.DisplayManagement.ModelBinding;
+using OrchardCore.Localization;
+using OrchardCore.Modules;
+using OrchardCore.Autoroute.Services;
+using OrchardCore.Autoroute.Model;
+using OrchardCore.Environment.Shell;
+
+namespace OrchardCore.ContentLocalization.Controllers
+{
+    [Feature("OrchardCore.ContentLocalization.ContentCulturePicker")]
+    public class ContentCulturePickerController : Controller, IUpdateModel
+    {
+        private readonly ILocalizationService _locationService;
+        private readonly IContentCulturePickerService _culturePickerService;
+        private readonly string _tenantPrefix;
+
+        public IHtmlLocalizer T { get; }
+
+        public ContentCulturePickerController(
+            IHtmlLocalizer<AdminController> localizer,
+            ILocalizationService locationService,
+            IContentCulturePickerService culturePickerService,
+            ShellSettings shellSettings
+           )
+        {
+            T = localizer;
+            _locationService = locationService;
+            _tenantPrefix = "/" + shellSettings.RequestUrlPrefix;
+            _culturePickerService = culturePickerService;
+        }
+
+        [HttpGet]
+        public async Task<IActionResult> RedirectToLocalizedContent(string targetCulture, string contentItemUrl)
+        {
+            if (string.IsNullOrEmpty(contentItemUrl))
+            {
+                return NotFound();
+            }
+
+            var supportedCultures = await _locationService.GetSupportedCulturesAsync();
+            if (!supportedCultures.Any(t => t == targetCulture))
+            {
+                return LocalRedirect(contentItemUrl);
+            }
+
+            // Redirect the user to the Content with the same localizationSet as the ContentItem of the current url
+            var contentItemId = await _culturePickerService.GetContentItemIdFromRoute(contentItemUrl);
+            if (!string.IsNullOrEmpty(contentItemId))
+            {
+                var contentItem = await _culturePickerService.GetRelatedContentItem(contentItemId, targetCulture);
+                var path = contentItem.As<AutoroutePart>()?.Path;
+                if (!string.IsNullOrEmpty(path))
```

Comment:
So I think here you can use something like
```
           var contentItemMetadata = await _contentManager.PopulateAspectAsync<ContentItemMetadata>(contentItem);
           var routes = contentItemMetadata.DisplayRouteValues;
```

to give you the content item routes without relying on the `AutoroutePart` as that is not the only way to route a content item (just the default one)